### PR TITLE
Enable multi-origin, same rpId use

### PR DESCRIPTION
**Pull request by reardencode**
https://github.com/apowers313/fido2-lib/issues/21

### CODE REVIEW

File: lib/validator.js [hunk 1]
```diff
@@ -209,13 +210,13 @@ async function validateId() {
...
 async function validateOrigin() {
-    var expectedOrigin = this.expectations.get("origin");
+    var expectedRpId = this.expectations.get("rpId");
     var clientDataOrigin = this.clientData.get("origin");
 
     var origin = checkOrigin(clientDataOrigin);
 
-    if (origin !== expectedOrigin) {
-        throw new Error("clientData origin did not match expected origin");
+    if (!new URL(origin).hostname.endsWith(expectedRpId)) {
```
Comment (davedoesdev) [inline lib/validator.js]: This is the relaxation mentioned in the spec? https://www.w3.org/TR/webauthn/#relying-party-identifier

**Comment by davedoesdev** [comment]
The current behavious _is_ wrong: If I set `rpId` to `example.com` for example and my site is on `foo.example.com` then the following error is thrown:

```
authnrData rpIdHash mismatch
```

because `validateRpIdHash()` uses the `origin` from `clientData` rather than expected `rpId`.

**Comment by davedoesdev** [comment]
I think we still need to check the origin though as the spec says:

> Verify that the value of C.origin matches the Relying Party's origin.

**Comment by davedoesdev** [comment]
So I think we need to modify `validateRpIdHash()` to check against the expected `rpId` and check the origin suffix is expected `rpId` (as well as equals expected `origin`).

We can make expected `rpId` optional so default is existing behaviour.

**Comment by JamesCullum** [comment]
Please make sure the tests pass and let me know once you are finished - last comments look like you are still working on it

**Comment by davedoesdev** [comment]
@JamesCullum I've abandoned fido2-lib in favour of https://github.com/davedoesdev/webauthn4js

**Comment by JamesCullum** [comment]
Sure, should I close this PR then or do you want to add the test and I merge it?

**Comment by davedoesdev** [comment]
Please close it, I won't be doing any more work on it

**Pull request closed without merging**